Initial scene support for Fibaro hubs

pbalogh77 opened the pull request:
Added initial support for scenes to fibaro hub integration

## Description:
Fibaro hub integration now imports and can activate scenes

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)
https://github.com/home-assistant/home-assistant.io/pull/7715

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

pbalogh77 committed:
Initial scene support

Added initial support for fibaro scenes

pbalogh77 committed:
removed comments

pbalogh77 committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into scenes

pbalogh77 commented on homeassistant/components/scene/fibaro.py:
```diff
@@ -0,0 +1,55 @@
+         for scene in hass.data[FIBARO_DEVICES]['scene']], True)
+
+
+class FibaroScene(FibaroDevice, Scene):
```
Yes, it's a unique id and the parent class of this (FibaroDevice) is retuning it as unique_id property. Should I just leave the entity_id attribute unfilled? Would that cause a breaking change or will I just end up with the same result anyway?

balloob replied:
You can safely remove the filling of the entity ID for all Fibaro components.

 - For new entities, it was never there, so it's fine
 - For existing entities, it was used as a suggestion when registering the entity ID, it is no longer used.

The nice thing is that now Home Assistant will start generating entity IDs based on name instead of unique ID, making it a more pleasant experience for the user.

pbalogh77 replied:
Removed it :-)

pbalogh77 commented on homeassistant/components/scene/fibaro.py:
```diff
@@ -0,0 +1,55 @@
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the scene."""
+        return self.fibaro_device.properties
```
It's a dict of all the attributes returned by fibaro, not sure what part of it would be useful. I could just as well remove it, scene support seems to be very very basic anyhow.

balloob replied:
That doesn't seem very useful

pbalogh77 replied:
removed it :-)

balloob commented:
Just a note that this is going to get a breaking change in the future as I want to break Home Assistant scenes out of the scene component and have a `scene_external` (with things like Fibaro) and `scene` which is scenes in Home Assistant land.

You're also welcome to open a PR to break up the component and then do a PR to add Fibaro to scene_external.

pbalogh77 committed:
cleanup based on code review

pbalogh77 committed:
Removed unused functions

pbalogh77 commented:
> Just a note that this is going to get a breaking change in the future as I want to break Home Assistant scenes out of the scene component and have a `scene_external` (with things like Fibaro) and `scene` which is scenes in Home Assistant land.
> 
> You're also welcome to open a PR to break up the component and then do a PR to add Fibaro to scene_external.

I'm way too junior in this yet to do that :-) However, when it's done, I'll migrate my code to the new component.

pbalogh77 committed:
grrr, my mistake.

My local pylint and flake8 are playing tricks with me

pbalogh77 committed:
Update homeassistant/components/scene/fibaro.py

MartinHjelmare commented on homeassistant/components/fibaro.py:
```diff
@@ -27,7 +27,8 @@
 
-FIBARO_COMPONENTS = ['binary_sensor', 'cover', 'light', 'sensor', 'switch']
+FIBARO_COMPONENTS = ['binary_sensor', 'cover', 'light',
+                     'sensor', 'switch', 'scene']
```
:abcd:

pbalogh77 replied:
ok :-)

MartinHjelmare commented on homeassistant/components/fibaro.py:
```diff
@@ -166,6 +169,22 @@ def _map_device_to_type(device):
+                room_name = 'Unknown'
+            else:
+                room_name = self._room_map[device.roomID].name
+            device.friendly_name = room_name + ' ' + device.name
```
Use new style string formatting instead of string concatenation.

MartinHjelmare commented on homeassistant/components/scene/fibaro.py:
```diff
@@ -0,0 +1,39 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
```
We could change this to `async_setup_platform` since we don't appear to do any I/O here.

MartinHjelmare commented on homeassistant/components/scene/fibaro.py:
```diff
@@ -0,0 +1,39 @@
+
+    async def async_activate(self):
+        """Activate the scene."""
+        self.fibaro_device.start()
```
Is this async safe, ie no I/O or blocking?

MartinHjelmare replied:
Unless it uses eg aiohttp, that probably isn't async safe.

MartinHjelmare replied:
Where's the source hosted? The github link on pypi is 404:
https://github.com/pbalogh/fiblary

MartinHjelmare replied:
It's this, right?
https://github.com/pbalogh77/fiblary

MartinHjelmare replied:
Ok, it's not async safe since requests is used in the library.

balloob replied:
Learn more about asyncio here: https://developers.home-assistant.io/docs/en/asyncio_101.html

pbalogh77 committed:
fixes based on code review

ABC ordered the list of platforms
changed setup platform to async
removed overloaded name property as the FibaroDevice parent class already provides this
Changed to new style string formatting

MartinHjelmare commented on homeassistant/components/scene/fibaro.py:
```diff
@@ -0,0 +1,35 @@
+class FibaroScene(FibaroDevice, Scene):
+    """Representation of a Fibaro scene entity."""
+
+    async def async_activate(self):
```
```suggestion
    def activate(self):
```

MartinHjelmare reviewed (commented):
That should be enough.

MartinHjelmare committed:
Update homeassistant/components/scene/fibaro.py

Co-Authored-By: pbalogh77 <peter.balogh2@gmail.com>

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Can be merged when build passes and docs PR is linked above in this PR description.

pbalogh77 commented:
Documentation added to https://github.com/home-assistant/home-assistant.io/pull/7715
